addons/hr_timesheet/models/hr_timesheet.py

```diff
@@ -270,42 +270,85 @@ def _timesheet_get_portal_domain(self):
-            vals['product_uom_id'] = analytic_account.company_id.project_time_mode_id.id
-        return vals
+        timesheet_indices = set()
+        task_ids, project_ids, account_ids = set(), set(), set()
+        for index, vals in enumerate(vals_list):
+            if not vals.get('project_id') and not vals.get('task_id'):
+                continue
+            timesheet_indices.add(index)
+            if vals.get('task_id'):
+                task_ids.add(vals['task_id'])
+            elif vals.get('project_id'):
+                project_ids.add(vals['project_id'])
+            if vals.get('account_id'):
+                account_ids.add(vals['account_id'])
+
+        task_per_id = {}
+        if task_ids:
+            tasks_read = self.env['project.task'].sudo().with_context(active_test=False).search_read(
+                domain=[('id', 'in', list(task_ids))],
+                fields=['project_id', 'company_id', 'analytic_account_id', 'partner_id'],
+                load=False,
+            )
+            for task_read in tasks_read:
+                task_per_id[task_read['id']] = task_read
+                if not task_read['project_id']:
+                    raise ValidationError(_('Timesheets cannot be created on a private task.'))
+                account_ids.add(task_read['analytic_account_id'])
+
+        project_per_id = {}
+        if project_ids:
+            projects_read = self.env['project.project'].sudo().with_context(active_test=False).search_read(
+                domain=[('id', 'in', list(project_ids))],
+                fields=['company_id', 'analytic_account_id', 'partner_id'],
+                load=False,
+            )
+            for project_read in projects_read:
+                project_per_id[project_read['id']] = project_read
+                account_ids.add(project_read['analytic_account_id'])
+
+        accounts_read = self.env['account.analytic.account'].sudo().with_context(active_test=False).search_read(
+            domain=[('id', 'in', list(account_ids))],
+            fields=['company_id', 'active'],
+            load=False,
+        )
+        account_per_id = {}
+        company_ids = set()
+        for account_read in accounts_read:
+            account_per_id[account_read['id']] = account_read
+            company_ids.add(account_read['company_id'])
+
+        uom_id_per_company_id = {
+            company_read['id']: company_read['project_time_mode_id']
+            for company_read in self.env['res.company'].sudo().search_read(
+                domain=[('id', 'in', list(company_ids))],
+                fields=['project_time_mode_id'],
+                load=False,
+            )
+        }
+
+        for index in timesheet_indices:
+            vals = vals_list[index]
+            data = task_per_id[vals['task_id']] if vals.get('task_id') else project_per_id[vals['project_id']]
+            if not vals.get('project_id'):
+                vals['project_id'] = data['project_id']
```

Comment:
If you do a browse then
```suggestion
                vals['project_id'] = data.project_id
```

Same idea below. 🙂